code/krepel/memory/construction.d

```diff
@@ -0,0 +1,172 @@
+  {
+    // TODO(Manu): Find out what the heck a xdtor is.
+
+    static if(Meta.HasMember!(Type, "__dtor"))  Instance.__dtor();
+    static if(Meta.HasMember!(Type, "__xdtor")) Instance.__dtor();
+    BlitInitialData(Instance[0 .. 1]);
+  }
+}
+
+void ConstructArray(Type, ArgTypes...)(Type[] Array, auto ref ArgTypes Args)
+{
+  static if(Meta.IsPlainOldData!Type)
+  {
+    static      if(Args.length == 0) BlitInitialData(Array);
+    else static if(Args.length == 1) Array[] = Args[0];
+    else
+    {
+      static assert(false,"Cannot initialize plain old data "
+                          "with more than one argument.");
+    }
+  }
+  else
+  {
+    static if(is(Type == class)) foreach(    Element; Array) Construct(Element);
+    else                         foreach(ref Element; Array) Construct(&Element);
+  }
+}
+
+
+void DestructArray(Type)(Type[] Array)
+{
+  static if(Meta.IsPlainOldData!Type)
+  {
+    BlitInitialData(Array);
+  }
+  else
+  {
+    static if(is(Type == class)) foreach(    Element; Array) Destruct(Element);
+    else                         foreach(ref Element; Array) Destruct(&Element);
+  }
+}
+
+private void BlitInitialData(Type)(Type[] BlitTargets)
+{
+  auto RawInit = cast(ubyte[])typeid(Type).initializer();
+
+  foreach(ref Target; BlitTargets)
+  {
+    static if(is(Type == class))
+      auto RawTarget = (cast(ubyte*) Target)[0 .. Meta.ClassInstanceSizeOf!Type];
+    else
+      auto RawTarget = (cast(ubyte*)&Target)[0 .. Type.sizeof];
+
+    if(RawInit.ptr) RawTarget[] = RawInit[];
+    else            RawTarget[] = 0;
+  }
+}
+
+
+//
+// Unit Tests
+//
+
+// Single struct object construction
+unittest
+{
+  nothrow @nogc static struct TestData
+  {
+    int Value;
+    float Precision;
+  }
+
+  ubyte[TestData.sizeof] Buffer = 0;
+  auto DataPtr = cast(TestData*)Buffer.ptr;
+  Construct(DataPtr, 42, 3.1415f);
+  assert(DataPtr.Value == 42);
+  assert(DataPtr.Precision == 3.1415f);
+
+  Destruct(DataPtr);
+  assert(DataPtr.Value != 42);
+  assert(DataPtr.Precision != 3.1415f);
+}
+
+// Single class object construction
+unittest
+{
+  class FooClass            {          int Data() { return 42;   } }
+  class BarClass : FooClass { override int Data() { return 1337; } }
+
+  void[1024] Buffer = void;
+  auto Bar = cast(BarClass)Buffer.ptr;
+  Construct!BarClass(Bar);
+  assert(Bar.Data == 1337);
+  FooClass Foo = Bar;
+  assert(Foo.Data == 1337);
+}
+
+// Array of structs
+unittest
+{
+  static struct TestData
+  {
+    int Value = 42;
+
+    ~this() { Value = 666; }
+  }
+
+  ubyte[5 * TestData.sizeof] Buffer;
+  const Size = TestData.sizeof;
+  auto Array = cast(TestData[])Buffer[0 .. 5 * TestData.sizeof];
+  foreach(ref Element; Array) assert(Element.Value == 0);
+  assert(Array[0].Value == 0);
+  ConstructArray(Array);
+  foreach(ref Element; Array)
+  {
+    assert(Element.Value == 42);
+    Element.Value = 1337;
+  }
+  DestructArray(Array);
+  foreach(ref Element; Array) assert(Element.Value == 42);
```

Comment:
Yea, I thought so as well... I've modelled the destruction logic after D's `destoy()` functions' behavior, which does not leave the object in an uninitialized state. It reverts it to the initial state `T.init`. I forgot to test that the destructor is actually called though.